src/lib/statistics/chunk_statistics/histograms/equal_element_count_histogram.cpp

```diff
@@ -0,0 +1,206 @@
+
+  if constexpr (std::is_same_v<T, std::string>) {
+    const auto [characters, prefix_length] =  // NOLINT (Extra space before [)
+        get_default_or_check_prefix_settings(supported_characters, string_prefix_length);
+    return std::make_shared<EqualElementCountHistogram<T>>(
+        bin_stats.mins, bin_stats.maxs, bin_stats.heights, bin_stats.distinct_count_per_bin,
+        bin_stats.bin_count_with_extra_value, characters, prefix_length);
+  } else {
+    DebugAssert(!static_cast<bool>(supported_characters) && !static_cast<bool>(string_prefix_length),
+                "Do not provide string prefix prefix arguments for non-string histograms.");
+    return std::make_shared<EqualElementCountHistogram<T>>(bin_stats.mins, bin_stats.maxs, bin_stats.heights,
+                                                           bin_stats.distinct_count_per_bin,
+                                                           bin_stats.bin_count_with_extra_value);
+  }
+}
+
+template <typename T>
+HistogramType EqualElementCountHistogram<T>::histogram_type() const {
+  return HistogramType::EqualElementCount;
+}
+
+template <typename T>
+BinID EqualElementCountHistogram<T>::bin_count() const {
+  return _heights.size();
+}
+
+template <typename T>
+BinID EqualElementCountHistogram<T>::_bin_for_value(const T value) const {
+  const auto it = std::lower_bound(_maxs.cbegin(), _maxs.cend(), value);
+  const auto index = static_cast<BinID>(std::distance(_maxs.cbegin(), it));
+
+  if (it == _maxs.cend() || value < _bin_min(index) || value > _bin_max(index)) {
+    return INVALID_BIN_ID;
+  }
+
+  return index;
+}
+
+template <typename T>
+BinID EqualElementCountHistogram<T>::_upper_bound_for_value(const T value) const {
+  const auto it = std::upper_bound(_maxs.cbegin(), _maxs.cend(), value);
+
+  if (it == _maxs.cend()) {
+    return INVALID_BIN_ID;
+  }
+
+  return static_cast<BinID>(std::distance(_maxs.cbegin(), it));
+}
+
+template <typename T>
+T EqualElementCountHistogram<T>::_bin_min(const BinID index) const {
+  DebugAssert(index < _mins.size(), "Index is not a valid bin.");
+  return _mins[index];
+}
+
+template <typename T>
+T EqualElementCountHistogram<T>::_bin_max(const BinID index) const {
+  DebugAssert(index < _maxs.size(), "Index is not a valid bin.");
+  return _maxs[index];
+}
+
+template <typename T>
+HistogramCountType EqualElementCountHistogram<T>::_bin_height(const BinID index) const {
+  DebugAssert(index < _heights.size(), "Index is not a valid bin.");
+  return _heights[index];
+}
+
+template <typename T>
+HistogramCountType EqualElementCountHistogram<T>::_bin_distinct_count(const BinID index) const {
+  DebugAssert(index < this->bin_count(), "Index is not a valid bin.");
+  return _distinct_count_per_bin + (index < _bin_count_with_extra_value ? 1 : 0);
+}
+
+template <typename T>
+HistogramCountType EqualElementCountHistogram<T>::total_count() const {
+  return std::accumulate(_heights.cbegin(), _heights.cend(), HistogramCountType{0});
```

Comment:
I suppose you explicity decided against storing this value explicitely? Because I'd expect `total_count()` to be called frequently because it effectively contains the non-null count of a segment, right?

Reply:
I did, for the sake of not storing redundant data that could, in theory, get contradictory.
It will be called for every estimation. I'm fine with storing it explicitly. Maybe someone else has an opinion on that?